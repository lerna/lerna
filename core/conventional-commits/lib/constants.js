"use strict";

// changelogs are always written with LF line endings
const EOL = "\n";

exports.EOL = EOL;

exports.BLANK_LINE = EOL + EOL;

exports.COMMIT_GUIDELINE =
  "See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.";

exports.CHANGELOG_HEADER_TEMPLATE = [
  "# Change Log",
  "",
  "All notable changes to this project will be documented in this file.",
];

exports.CHANGELOG_HEADER = [...exports.CHANGELOG_HEADER_TEMPLATE, exports.COMMIT_GUIDELINE].join(EOL);
