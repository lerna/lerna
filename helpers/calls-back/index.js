"use strict";

module.exports = callsBack;

// for mocking the behaviour of methods that accept a callback
function callsBack(err, val) {
  return (...args) => process.nextTick(args.pop(), err, val);
}
