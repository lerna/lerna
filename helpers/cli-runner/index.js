"use strict";

const execa = require("execa");

const LERNA_BIN = require.resolve("../../core/lerna/cli");

module.exports = runner;

function runner(cwd, env) {
  const opts = {
    cwd,
    env: Object.assign(
      {
        CI: "true",
        // always turn off chalk
        FORCE_COLOR: "0",
      },
      env
    ),
    // when debugging integration test snapshots, uncomment next line
    // stdio: ["ignore", "inherit", "inherit"],
  };

  return (...args) => execa("node", [LERNA_BIN].concat(args), opts);
}
