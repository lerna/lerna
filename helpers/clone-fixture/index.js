"use strict";

const execa = require("execa");
const fileUrl = require("file-url");
const tempy = require("tempy");
const initFactory = require("@lerna-test/init-fixture");

module.exports = cloneFixture;

function cloneFixture(startDir) {
  const initFixture = initFactory(startDir);

  return (...args) =>
    initFixture(...args).then((cwd) => {
      const repoDir = tempy.directory();
      const repoUrl = fileUrl(repoDir, { resolve: false });

      return execa("git", ["init", "--bare"], { cwd: repoDir })
        .then(() => execa("git", ["checkout", "-B", "main"], { cwd }))
        .then(() => execa("git", ["remote", "add", "origin", repoUrl], { cwd }))
        .then(() => execa("git", ["push", "-u", "origin", "main"], { cwd }))
        .then(() => ({
          cwd,
          repository: repoUrl,
        }));
    });
}
