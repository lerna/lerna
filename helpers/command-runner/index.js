"use strict";

const yargs = require("yargs/yargs");
const globalOptions = require("@lerna/global-options");

module.exports = commandRunner;

/**
 * A higher-order function to help with passing _actual_ yargs-parsed argv
 * into command constructors (instead of artificial direct parameters).
 *
 * @param {Object} commandModule The yargs command exports
 * @return {Function} with partially-applied yargs config
 */
function commandRunner(commandModule) {
  const cmd = commandModule.command.split(" ")[0];

  return cwd => {
    // create a _new_ yargs instance every time cwd changes to avoid singleton pollution
    const cli = yargs([], cwd)
      .strict()
      .exitProcess(false)
      .detectLocale(false)
      .showHelpOnFail(false)
      .wrap(null)
      .command(commandModule);

    globalOptions(cli);

    return (...args) =>
      new Promise((resolve, reject) => {
        const yargsMeta = {};

        const context = {
          cwd,
          onResolved: result => {
            // success does not resolve anything, currently
            resolve(Object.assign({}, result, yargsMeta));
          },
          onRejected: result => {
            Object.assign(result, yargsMeta);
            // tests expect errors thrown to indicate failure,
            // _not_ just non-zero exitCode
            reject(result);
          },
        };

        const parseFn = (yargsError, parsedArgv, yargsOutput) => {
          // this is synchronous, before the async handlers resolve
          Object.assign(yargsMeta, { parsedArgv, yargsOutput });
        };

        cli
          .fail((msg, err) => {
            // since yargs 10.1.0, this is the only way to catch handler rejection
            // _and_ yargs validation exceptions when using async command handlers
            const actual = err || new Error(msg);
            // backfill exitCode for test convenience
            yargsMeta.exitCode = "exitCode" in actual ? actual.exitCode : 1;
            context.onRejected(actual);
          })
          .parse([cmd, ...args], context, parseFn);
      });
  };
}
