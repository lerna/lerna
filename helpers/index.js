"use strict";

// eslint-disable-next-line import/no-extraneous-dependencies, node/no-extraneous-require
const { Project } = require("@lerna/project");
const normalizeNewline = require("normalize-newline");
const path = require("path");
const normalizePath = require("normalize-path");

const git = require("./git");
const fixtures = require("./fixtures");
const npm = require("./npm");
const cli = require("./cli");

module.exports = {
  ...git,
  ...fixtures,
  ...npm,
  ...cli,
  updateLernaConfig,
  multiLineTrimRight,
  normalizeRelativeDir,
};

/**
 * Update lerna config inside a test case.
 *
 * @param {String} testDir where target lerna.json exists
 * @param {Object} updates mixed into existing JSON via Object.assign
 */
function updateLernaConfig(testDir, updates) {
  const project = new Project(testDir);

  Object.assign(project.config, updates);

  return project.serializeConfig();
}

function multiLineTrimRight(str) {
  return normalizeNewline(str)
    .split("\n")
    .map((line) => line.trimRight())
    .join("\n");
}

function normalizeRelativeDir(testDir, filePath) {
  return normalizePath(path.relative(testDir, filePath));
}
