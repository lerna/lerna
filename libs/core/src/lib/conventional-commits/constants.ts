export type VersioningStrategy = "independent" | "fixed";
export type ChangelogType = "fixed" | "independent" | "root";
export type ChangelogPresetConfig = string | { name: string; [key: string]: unknown };
export type BaseChangelogOptions = {
  changelogPreset?: ChangelogPresetConfig;
  rootPath: string;
  tagPrefix?: string;
  conventionalBumpPrerelease?: boolean;
};

// changelogs are always written with LF line endings
export const EOL = "\n";

export const BLANK_LINE = EOL + EOL;

export const COMMIT_GUIDELINE =
  "See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.";

export const CHANGELOG_HEADER = [
  "# Change Log",
  "",
  "All notable changes to this project will be documented in this file.",
  COMMIT_GUIDELINE,
].join(EOL);
