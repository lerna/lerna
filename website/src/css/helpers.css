.link--inset {
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}
