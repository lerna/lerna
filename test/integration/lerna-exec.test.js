import execa from "execa";

import { LERNA_BIN } from "../helpers/constants";
import initFixture from "../helpers/initFixture";

describe("lerna exec", () => {
  test.concurrent("--ignore <pkg> ls -- -1", async () => {
    const cwd = await initFixture("ExecCommand/basic");
    const args = [
      "exec",
      "--ignore=package-1",
      "ls",
      "--concurrency=1",
      "--",
      // args to ls
      "-1",
    ];

    const { stdout, stderr } = await execa(LERNA_BIN, args, { cwd });
    expect(stdout).toMatchSnapshot("stdout: ls --ignore");
    expect(stderr).toMatchSnapshot("stderr: ls --ignore");
  });

  test.concurrent("ls --scope <pkg>", async () => {
    const cwd = await initFixture("ExecCommand/basic");
    const args = [
      "exec",
      "--concurrency=1",
      "ls",
      "--scope=package-1",
      // no args to ls
    ];

    const { stdout, stderr } = await execa(LERNA_BIN, args, { cwd });
    expect(stdout).toMatchSnapshot("stdout: ls --scope");
    expect(stderr).toMatchSnapshot("stderr: ls --scope");
  });

  test.concurrent("without --", async () => {
    const cwd = await initFixture("ExecCommand/basic");
    const args = [
      "--concurrency=1",
      "exec",
      "ls",
      // no --
      "-C",
    ];

    const { stdout, stderr } = await execa(LERNA_BIN, args, { cwd });
    expect(stdout).toMatchSnapshot("stdout: without --");
    expect(stderr).toMatchSnapshot("stderr: without --");
  });

  test.concurrent("echo $LERNA_PACKAGE_NAME", async () => {
    const cwd = await initFixture("ExecCommand/basic");
    const args = [
      "exec",
      "--concurrency=1",
      "echo",
      (process.platform == "win32" ? "%LERNA_PACKAGE_NAME%" : "$LERNA_PACKAGE_NAME"),
    ];

    const { stdout, stderr } = await execa(LERNA_BIN, args, { cwd });
    expect(stdout).toMatchSnapshot("stdout: echo LERNA_PACKAGE_NAME");
    expect(stderr).toMatchSnapshot("stderr: echo LERNA_PACKAGE_NAME");
  });

  test.concurrent("<cmd> --parallel", async () => {
    const cwd = await initFixture("ExecCommand/basic");
    const args = [
      "exec",
      "ls",
      "--parallel",
      // no --
      "-C",
    ];

    const { stdout, stderr } = await execa(LERNA_BIN, args, { cwd });
    expect(stderr).toMatchSnapshot("stderr: <cmd> --parallel");

    // order is non-deterministic, so assert individually
    expect(stdout).toMatch("package-1: file-1.js");
    expect(stdout).toMatch("package-1: package.json");
    expect(stdout).toMatch("package-2: file-2.js");
    expect(stdout).toMatch("package-2: package.json");
  });

  test.concurrent("--parallel <cmd>", async () => {
    const cwd = await initFixture("ExecCommand/basic");
    const args = [
      "exec",
      "--parallel",
      "ls",
      // no --
      "-C",
    ];

    const { stdout, stderr } = await execa(LERNA_BIN, args, { cwd });
    expect(stderr).toMatchSnapshot("stderr: --parallel <cmd>");

    // order is non-deterministic, so assert individually
    expect(stdout).toMatch("package-1: file-1.js");
    expect(stdout).toMatch("package-1: package.json");
    expect(stdout).toMatch("package-2: file-2.js");
    expect(stdout).toMatch("package-2: package.json");
  });
});
