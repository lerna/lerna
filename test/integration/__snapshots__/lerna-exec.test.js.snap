// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stderr: <cmd> --parallel 1`] = `
"lerna info version __TEST_VERSION__
lerna info exec in 2 package(s): ls"
`;

exports[`stderr: --parallel <cmd> 1`] = `
"lerna info version __TEST_VERSION__
lerna info exec in 2 package(s): ls"
`;

exports[`stderr: echo LERNA_PACKAGE_NAME 1`] = `"lerna info version __TEST_VERSION__"`;

exports[`stderr: ls --ignore 1`] = `
"lerna info version __TEST_VERSION__
lerna info ignore package-1"
`;

exports[`stderr: ls --scope 1`] = `
"lerna info version __TEST_VERSION__
lerna info scope package-1"
`;

exports[`stderr: without -- 1`] = `"lerna info version __TEST_VERSION__"`;

exports[`stdout: echo LERNA_PACKAGE_NAME 1`] = `
"package-1
package-2"
`;

exports[`stdout: ls --ignore 1`] = `
"file-2.js
package.json"
`;

exports[`stdout: ls --scope 1`] = `
"file-1.js
package.json"
`;

exports[`stdout: without -- 1`] = `
"file-1.js
package.json
file-2.js
package.json"
`;
